pkg/recommender.py

```diff
@@ -145,6 +147,19 @@ def dispatch(self, speaker_id, evt):
         if not isinstance(evt, np.ndarray):
             evt = np.array(evt)
 
+        #safety in case of numpy array or float
+        speaker_id = int(speaker_id)
+
+        #system must be initialized (no cooldown)
+        if not INITIALIZED_SCHED_EVT:
+            log('Scheduled events not yet initialized')
+            return
+
+        #do not interuprt scheduled events messages (no cooldown)
+        if IN_PROGRESS_EVT:
+            log('Scheduled events messages are in progress')
```

Comment:
What is this for? We receive another event when one is still in progress and cooldown has expired?

Reply:
yeah, I added this so a recommendation does not interrupt scheduled events (technically only morning messages). I can remove it if you prefer, but that will leave us vulnerable for an interruption.

Reply:
ok, just curious then, why two recommendations won't have such problem? We can ensure the previous have completed before the cooldown expires? Or we just let the new one overwrite the previous one?

anyway, it should not be a global variable, right? it means the state of this recommender instance. Let's move it as a class variable under, sth like `self.schedule_evt_in_progress`.  

btw, did we implement the logic to ignore events at night (after end time and before start time)?

Reply:
When we send out a recommendation, we start our cooldown period. So the second recommendation will be blocked by the cooldown. However, when we send a scheduled event, we do not start a cooldown period. We also do not check if there is a cooldown period active (we will always send a scheduled event at the scheduled time). Should I change this logic?

Yes, the logic to ignore events at night is on line 203:
 https://github.com/aobo-y/caregiver_recomm/pull/28/files#diff-fc9fb584240db0bb6e88ac0b15409b70R203

and the logic to ignore events after we have sent the maximum amount of messages is on line 194:
https://github.com/aobo-y/caregiver_recomm/pull/28/files#diff-fc9fb584240db0bb6e88ac0b15409b70R194

For both of these cases, a cooldown period is initialized even if we have ignored the event. Should I put this logic in the `dispatch` function instead of the `_process_evt` function? If I put this logic in the dispatch function a cooldown period will not be initialized. 

Also for both of these cases I am also using global variables such as `MESSAGES_SENT_TODAY`, `MAX_MESSAGES`, `TIME_MORN_DELT`, and `TIME_EV_DELT`. Would it be better to make these class variables too?

Reply:
Oh to finish answering your questions, right now we _cannot_ ensure that the previous recommendation messages have been completed. We just hope that they are completed within the cooldown period. However, since we are now adding a larger poll time between unanswered questions (like discussed on Friday), I think it is very possible that recommendations will be overwritten :(

I asked Dr. Wang about this on Friday and he suggested just to make a longer cooldown period depending on if we sleep 30 or 60 minutes after a recommendation. Would it be better to add logic to ensure that recommendations are completed?

**sorry for the long messages

Reply:
sure, for the cooldown, just keep it. Let the overwrite be the default behaviour and we can enlarge the number as Prof. Wang suggested. The simple the better.

For the schedule event & recommendation conflicts, do you think we can try to simplify the logic?
Say we just have one variable `self.recomm_start`, whose default is `false`. In the `dispatch`, we check its value at first. Set it to `true` after morning message complete, like ur current `in_progress`. Set it back to `false` when evening message start. Seems this may unify your logic of `in_progress` and checking start & end time. Since I haven't gone through all the details, I may be wrong. Please let me know if this is a good idea and if you have any concerns.

Reply:
yes, stuff we read from db should not be a global variable

Reply:
yes this is a good idea, I will make the changes!